### ISSUE TITLE
Implement generic irregular smoothing

### ISSUE THREAD

**Issue by eackermann** [nelpy/nelpy#291]
Savva has been working on this for SpikeTrainArrays, but I'm parking it here as a generally useful feature that we should support as generically and as efficiently as possible.

Ideally, I want it to be as easy as

```python
all_spikes = spikes.flatten()  # collapse all units into a multiunit spike train
mua = all_spikes.bin(dt=0.02, smooth=True, sigma=0.01)
```
to compute multiunit activity from a `SpikeTrainArray` where we have smoothed with a Gaussian kernel of 10 ms, and binned / discretized the MUA into 20 ms bins.

We want to be able to smooth (i) within and (ii) across epochs, and we want to be able to specify whether we want to have the smoothing be (a) mass-preserving, or (b) averaging. There are other options, too, but these are the most prominent ones AFAIK.

The irregular smoothing should be implemented generically, so that we can smooth an irregular `AnalogSignalArray`, for example, or 2D position estimates, or whatever.

This is closely related to Gaussian process regression, and we may especially want to look into that more formally for 2D or ND surface estimation, as it also relates to firing rate estimation in higher dimensional spaces.

We need to get from Savva what he has already done, and we need to bring in James for his GP work, and we need to think carefully about how to organize such a feature to make it generally useful.

Related (but old) issues include #171 #148 #156 and #241

**Comment by eackermann** [comment]
See also https://github.com/nelpy/example-analyses/blob/master/SpikeTrainSmoothing.ipynb

**Comment by jchutrue** [comment]
One question I have, specifically for spike train smoothing. As an example, suppose we have a spike at 0.5 s with bin edges given by 0.0, 0.2, 0.4, 0.6, 0.8, 1.0 s. 

Assume that for smoothing, we want `sigma=0.1` and the kernel to be truncated at `4*sigma`. Do we evaluate the Gaussian kernel only at the bin centers?

If so, the effective bandwidth changes depending on the spike times. Suppose the spike is at 0.55 s instead. Then the lower bound is 0.55 - 4*0.1 = 0.15 s. The closest bin center is at 0.1, but that is below the lower bound of 0.15 s, so we take the neighboring bin center of 0.3 s. However, this is only (0.3 - 0.55)/0.1 = 2.5*sigma.

My opinion is that we should evaluate the Gaussian kernel more densely than just the bin centers. However, this will also slow down execution time because each spike would need to be processed one-by-one.

**Comment by eackermann** [comment]
The idea was always to center each spike's Gaussian on the spike event, not on some bin center. But there are fast ways of doing that, and we don't need to loop over all spikes. The notebook I had shared a while back also demonstrated that the kernels are centered around each spike event, and that we could easily vectorize such computations. Savva improved on my early notebook quite a bit by adding more functionality, but it was never completed.

**Comment by jchutrue** [comment]
I agree that computing the Gaussian centered on each spike event can be vectorized. However, I am arguing that it is not easy to assign the results of the computation to the correct bins in a vectorized manner. How would you do it for the situation I described above?
